core/src/worker/workflow/managed_run.rs

```diff
@@ -0,0 +1,602 @@
+
+            let immediate_resolutions = (self.local_activity_request_sink)(new_local_acts);
+            for resolution in immediate_resolutions {
+                self.wfm
+                    .notify_of_local_result(LocalResolution::LocalActivity(resolution))?;
+            }
+
+            let data = CompletionDataForWFT {
+                task_token: completion.task_token,
+                query_responses: completion.query_responses,
+                has_pending_query: completion.has_pending_query,
+                activation_was_only_eviction: completion.activation_was_only_eviction,
+            };
+            if self.wfm.machines.outstanding_local_activity_count() == 0 {
+                Ok((None, data, self))
+            } else {
+                let wft_timeout: Duration = self
+                    .wfm
+                    .machines
+                    .get_started_info()
+                    .and_then(|attrs| attrs.workflow_task_timeout)
+                    .ok_or_else(|| {
+                        WFMachinesError::Fatal(
+                            "Workflow's start attribs were missing a well formed task timeout"
+                                .to_string(),
+                        )
+                    })?;
+                let heartbeat_tx = heartbeat_tx.clone();
+                Ok((
+                    Some((heartbeat_tx, completion.start_time, wft_timeout)),
+                    data,
+                    self,
+                ))
+            }
+        }
+        .await;
+
+        match outcome {
+            Ok((None, data, me)) => {
+                me.finish_completion(resp_chan, data, false);
+                Ok(())
+            }
+            Ok((Some((chan, start_t, wft_timeout)), data, me)) => {
+                if let Some(wola) = me.waiting_on_la.as_mut() {
+                    wola.heartbeat_timeout_task.abort();
+                }
+                me.waiting_on_la = Some(WaitingOnLAs {
+                    wft_timeout,
+                    completion_dat: Some((data, resp_chan)),
+                    hb_chan: chan.clone(),
+                    heartbeat_timeout_task: start_heartbeat_timeout_task(
+                        chan,
+                        start_t,
+                        wft_timeout,
+                    ),
+                });
+                Ok(())
+            }
+            Err(e) => Err(RunUpdateErr {
+                source: e,
+                complete_resp_chan: Some(resp_chan),
+            }),
+        }
+    }
+
+    async fn check_more_work(
+        &mut self,
+        want_to_evict: Option<RequestEvictMsg>,
+        has_pending_queries: bool,
+    ) -> Result<Option<ActivationOrAuto>, RunUpdateErr> {
+        if self.wfm.machines.has_pending_jobs() && !self.am_broken {
+            Ok(Some(ActivationOrAuto::LangActivation(
+                self.wfm.get_next_activation().await?,
+            )))
+        } else {
+            if has_pending_queries && !self.am_broken {
```

Comment:
Does this mean we can't query failed workflows? (not 100% sure what it means to be broken)

Reply:
Can't query workflows who've had some internal error. Failing them the normal way is fine.